pkg/conversion/conversion_rules.go

```diff
@@ -0,0 +1,208 @@
+	Program cel.Program
+}
+
+// CompileConversions compiles conversion rules.
+func CompileConversions(
+	apiConversion *apisv1alpha1.APIConversion,
+	structuralSchemas map[string]*structuralschema.Structural,
+) (map[string][]*CompiledRule, error) {
+	compiledRules := make(map[string][]*CompiledRule)
+	var errs []error
+
+	for i := range apiConversion.Spec.Conversions {
+		c := apiConversion.Spec.Conversions[i]
+		path := field.NewPath("spec", "conversions").Index(i)
+		rulesForVersion, err := compileConversion(path, &c, structuralSchemas)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		compiledRules[c.From] = rulesForVersion
+	}
+
+	return compiledRules, kerrors.NewAggregate(errs)
+}
+
+func compileConversion(path *field.Path, c *apisv1alpha1.APIVersionConversion, structuralSchemas map[string]*structuralschema.Structural) ([]*CompiledRule, error) {
+	rulesForVersion := make([]*CompiledRule, 0, len(c.Rules))
+
+	schema, exists := structuralSchemas[c.From]
+	if !exists {
+		return nil, field.Invalid(path.Child("from"), c.From, "unable to find structural schema for version")
+	}
+
+	var errs []error
+	for j, rule := range c.Rules {
+		rulePath := path.Child("rules").Index(j)
+		cr, err := compileRule(rulePath, rule, schema)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		rulesForVersion = append(rulesForVersion, cr)
+	}
+
+	return rulesForVersion, kerrors.NewAggregate(errs)
+}
+
+func compileRule(rulePath *field.Path, rule apisv1alpha1.APIConversionRule, schema *structuralschema.Structural) (*CompiledRule, error) {
+	cr := &CompiledRule{
+		FromPath: rule.Field,
+		ToPath:   rule.Destination,
+	}
+
+	ruleField := rule.Field
+	if ruleField[0] == '.' {
+		ruleField = ruleField[1:]
+	}
+	schemaForField, err := schemaIndexByJSONPath(schema, ruleField)
+	if err != nil {
+		return nil, field.Invalid(rulePath.Child("field"), rule.Field, err.Error())
+	}
+
+	if rule.Transformation == "" {
+		return cr, nil
+	}
+
+	env, err := getCELEnv(schemaForField)
+	if env == nil {
+		err = errors.New("environment was nil")
+	}
+	if err != nil {
+		return nil, field.InternalError(rulePath.Child("field"), fmt.Errorf("error creating CEL environment: %w", err))
+	}
+
+	ast, issues := env.Compile(rule.Transformation)
+	if issues != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error compiling CEL program: %v", issues.Err()))
+	}
+
+	program, err := env.Program(ast,
+		cel.EvalOptions(cel.OptOptimize),
+		cel.OptimizeRegex(library.ExtensionLibRegexOptimizations...),
+		cel.InterruptCheckFrequency(celCheckFrequency),
+	)
+	if err != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error creating CEL program: %v", err))
+	}
+
+	cr.Program = program
+
+	return cr, nil
+}
+
+func getCELEnv(structuralSchema *structuralschema.Structural) (*cel.Env, error) {
```

Comment:
merely `createCELEnv`. This is not a getter.

Reply:
Fixing